vcpkg/src/lib.rs

```diff
@@ -398,19 +398,196 @@ fn find_vcpkg_target(cfg: &Config, target_triplet: &TargetTriplet) -> Result<Vcp
+                                "lib"
+                            } else {
+                                ""
+                            },
+                            lib_flag.trim_left_matches("-l"),
+                            target_triplet.lib_suffix
+                        );
+                        libs.push(lib);
+                    }
+                }
+            }
+        }
+
+        Ok(PcFile {
+            id: id.to_string(),
+            libs: libs,
+            deps: deps,
+        })
+    }
+}
+
+/// Collection of PcFile.  Can be built and queried as a set of .pc files.
+#[derive(Debug)]
+struct PcFiles {
+    files: BTreeMap<String, PcFile>,
+}
+impl PcFiles {
+    fn load_pkgconfig_dir(vcpkg_target: &VcpkgTarget, path: &PathBuf) -> Result<PcFiles, Error> {
+        let mut files = BTreeMap::new();
+        for dir_entry in try!(path.read_dir().map_err(|e| {
+            Error::VcpkgInstallation(format!(
+                "Missing pkgconfig directory {}: {}",
+                path.to_string_lossy(),
+                e
+            ))
+        })) {
+            let dir_entry = try!(dir_entry.map_err(|e| {
+                Error::VcpkgInstallation(format!(
+                    "Troubling reading pkgconfig dir {}: {}",
+                    path.to_string_lossy(),
+                    e
+                ))
+            }));
+            // Only look at .pc files.
+            if dir_entry.path().extension() != Some(OsStr::new("pc")) {
+                continue;
+            }
+            let pc_file = try!(PcFile::parse_pc_file(vcpkg_target, &dir_entry.path()));
+            files.insert(pc_file.id.to_owned(), pc_file);
+        }
+        Ok(PcFiles { files: files })
+    }
+    /// Use the .pc files as a hint to the library sort order.
+    fn fix_ordering(&self, mut libs: Vec<String>) -> Vec<String> {
+        // Overall heuristic: for each library given as input, identify which PcFile declared it.
+        // Then, looking at that PcFile, check its Requires: (deps), and if the pc file for that
+        // dep is in our set, check if its libraries are in our set of libs.  If so, move it to the
+        // end to ensure it gets linked afterwards.
+
+        // We may need to do this a few times to properly handle the case where A -> (depends on) B
+        // -> C -> D and libraries were originally sorted D, C, B, A.  Avoid recursion so we don't
+        // have to detect potential cycles.
+        let mut tries_left = 3;
+        loop {
+            let mut required_lib_order: Vec<String> = Vec::new();
+            for lib in &libs {
+                required_lib_order.push(lib.to_owned());
+                if let Some(pc_file) = self.locate_pc_file_by_lib(lib) {
+                    // Consider its requirements:
+                    for dep in &pc_file.deps {
+                        // Only consider pkgconfig dependencies we know about.
+                        let dep_pc_file = match self.files.get(dep) {
```

Comment:
Here I might write this as an `if let Some(dep_pc_file) = self.files.get(dep) { ... }`.